### ISSUE TITLE
Fields included via $allOf into AsyncAPI messages are not constrained correctly

### ISSUE THREAD

**Issue by DavidGregory084** [asyncapi/modelina#1611]
#### Describe the bug

I have some shared fields that are used in multiple messages in my schema:

```yaml
components:
  schemas:
    AbstractData:
      description: Abstract fields
      required:
        - capturedAt
        - deviceId
      properties:
        deviceId:
          type: string
          format: uuid
        capturedAt:
          type: string
          format: date-time
```

When these are included into my message...

```yaml
  messages:
    ConcreteData:
      description: Concrete implementation as a message
      payload:
        x-parser-schema-id: ConcreteData
        type: object
        additionalProperties: false
        allOf:
          - $ref: "#/components/schemas/AbstractData"
        required:
          - someValue
        properties:
          someValue:
            type: number
```

then the generated code does not constrain the fields correctly:

```kotlin
data class ConcreteData(
    val someValue: Double,
    val deviceId: String,
    val capturedAt: String,
)
```

If I comment out the `deviceId` field, then the `capturedAt` field is represented correctly as an `OffsetDateTime`:

```kotlin
data class ConcreteData(
    val someValue: Double,
    val capturedAt: java.time.OffsetDateTime,
)
```

#### How to Reproduce

Please see [this Codesandbox](https://codesandbox.io/p/sandbox/gallant-meadow-3lk6p8?file=%2Findex.js%3A18%2C13&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clp05l7a900072e68mt71nzt1%2522%252C%2522sizes%2522%253A%255B45.17401392%252C54.82598608%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clp05l7a800032e687lgskhf4%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clp05l7a900052e68duj9fqon%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clp05l7a900062e68bm48mp4d%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B100%252C0%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clp05l7a800032e687lgskhf4%2522%253A%257B%2522id%2522%253A%2522clp05l7a800032e687lgskhf4%2522%252C%2522activeTabId%2522%253A%2522clp05n5y400fm2e68imqj10f2%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clp05l7a800022e68dqjc8dsb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522id%2522%253A%2522clp05lvhc00722e68et3kfl1t%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Findex.js%2522%252C%2522id%2522%253A%2522clp05n5y400fm2e68imqj10f2%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftasks.json%2522%252C%2522id%2522%253A%2522clp06snts01zs2e68689guw7y%2522%252C%2522mode%2522%253A%2522temporary%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522clp05l7a900062e68bm48mp4d%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clp05l7a900062e68bm48mp4d%2522%257D%252C%2522clp05l7a900052e68duj9fqon%2522%253A%257B%2522id%2522%253A%2522clp05l7a900052e68duj9fqon%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clp05l7a800042e68hzt8euj0%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clp05l7a800042e68hzt8euj0%2522%257D%257D%252C%2522showDevtools%2522%253Afalse%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A22.604166666666657%257D) or [gist](https://gist.github.com/DavidGregory084/f91e3e0dcba2702c3c91c8a4996e0a17).

#### Expected behavior

I expect that fields will be constrained according to their `format`, rather than a single field's constraints being applied to every field of that type regardless of the `format` option. I was attempting to use `typeMapping` to set custom types for these fields when I noticed this issue:

```js
  typeMapping: {
    String: (opts) => {
      const { constrainedModel } = opts;
      switch (constrainedModel.options.format) {
        case "date-time": {
          return "java.time.Instant";
        }
        case "uuid": {
          return "java.util.UUID";
        }
        default: {
          return KotlinDefaultTypeMapping.String(opts);
        }
      }
    },
  },
```

**Comment by kennethaasan** [comment]
Hi, I was trying to reproduce your issue, but it seems fine here: https://github.com/asyncapi/modelina/pull/1614/files
Would you be able to reproduce this problem in a test?

**Comment by DavidGregory084** [comment]
@kennethaasan there is a link above to a CodeSandbox which reproduces the issue:
<img width="557" alt="image" src="https://github.com/asyncapi/modelina/assets/2992938/bfb9b0c7-6210-4146-859f-8901e4238fb3">

Perhaps it's been fixed on `main`? That would be great if so!

**Comment by DavidGregory084** [comment]
There is a difference between your test and the example I provided - the use of `@asyncapi/parser` to parse the doc. Perhaps this issue actually belongs there?

**Comment by kennethaasan** [comment]
Sounds like something for @jonaslagoni to take a look at

**Comment by DavidGregory084** [comment]
Hmm I was able to replicate the issue with [this commit](https://gist.github.com/DavidGregory084/f91e3e0dcba2702c3c91c8a4996e0a17#file-0001-use-a-schema-parsed-with-asyncapi-parser-to-replicat-patch), however while experimenting with it I realised that the issue is triggered by a missing `type: object` declaration in the `AbstractData` schema.

I was surprised that none of the tooling told me about this, but apparently a schema without a `type` is still valid, leading to nonsensical results like this:

<img width="1603" alt="image" src="https://github.com/asyncapi/modelina/assets/2992938/043bda9d-46b9-4856-8c9c-1c0b99d9c808">

See [this issue](https://github.com/json-schema/json-schema/issues/172) for more details.

Let's close this with lesson learned: don't forget the `type` keyword if you're afraid of nasal demons.
